Add copilot instructions for creating OpenFGA models (#61)
openfga/sample-stores: .github/copilot-instructions.md

<!-- Use this file to provide workspace-specific custom instructions to Copilot. For more details, visit https://code.visualstudio.com/docs/copilot/copilot-customization#_use-a-githubcopilotinstructionsmd-file -->

# **Authoring OpenFGA Models**

This guide provides a comprehensive overview of authoring OpenFGA authorization models. It covers core concepts, the modeling language, relationship definitions, and testing methodologies, drawing insights from the openfga/sample-stores repository for practical examples.

## **1. Introduction to OpenFGA and Authorization Modeling**

OpenFGA is an open-source authorization solution that empowers developers to implement fine-grained access control within their applications through an intuitive modeling language.

It functions as a flexible authorization engine, simplifying the process of defining application permissions.

> Inspired by Google's Zanzibar paper, OpenFGA primarily champions Relationship-Based Access Control (ReBAC), while also effectively addressing use cases for Role-Based Access Control (RBAC) and Attribute-Based Access Control (ABAC).1 Its "developer-first" philosophy is evident in its Domain Specific Language (DSL) and supporting tools, which lower the barrier to entry for developers.

Reviewer: ```suggestion
Inspired by Google's Zanzibar paper, OpenFGA primarily champions Relationship-Based Access Control (ReBAC), while also effectively addressing use cases for Role-Based Access Control (RBAC) and Attribute-Based Access Control (ABAC). Its "developer-first" philosophy is evident in its Domain Specific Language (DSL) and supporting tools, which lower the barrier to entry for developers.
```

The core purpose of an authorization model is to define a system's permission structure, answering questions like, "Can user U perform action A on object O?". By externalizing authorization logic from application code, OpenFGA provides a robust mechanism for managing complex access policies, especially in large-scale systems. The modeling language is designed to be powerful for engineers yet accessible to other team stakeholders, fostering collaborative policy development.

## **2. OpenFGA Core Concepts: The Building Blocks of Your Model**

To effectively model authorization in OpenFGA, it is essential to understand its core building blocks:

* **Authorization Model:** A static blueprint that combines one or more type definitions to precisely define the permission structure of a system. It represents the  

>   *possible* relations between users and objects. Models are generally stable and immutable; each modification creates a new version with a unique ID.

Reviewer: ```suggestion
  *possible* relations between users and objects. Models are immutable; each modification creates a new version with a unique ID. Models aren't expected change often - only when new product features or change in functionality is introduced. They're also generally expected to be backward compatible, but can break backward compatibility once the system has completely moved off the older relations in it.
```

* **Type:** A string that defines a class of objects sharing similar characteristics (e.g., user, document, folder, organization, team, repo).
* **Object:** A specific instance of a defined Type (e.g., `document:roadmap`, `user:anne`, `organization:acme`). An object's relationships are defined through relationship tuples and the authorization model.

> * **User:** An entity that can be related to an object. A user can be a specific individual (e.g., user:anne), a wildcard representing everyone (`user:*`), or a userset (e.g., `team:product#member`), which denotes a group of users.  

Reviewer: ```suggestion
* **User:** An entity that can be related to an object. A user can be a specific individual (e.g., user:anne), a wildcard representing everyone of a certain type (e.g. `user:*` which means all users), or a userset (e.g., `team:product#member`), which denotes a group of users.  
```

> * **Relation:** A string defined within a type definition in the authorization model. It specifies the *possibility* of a relationship existing between an object of that type and a user. Relation names are arbitrary (e.g., owner, editor, viewer, member, admin).

Reviewer: ```suggestion
* **Relation:** A string defined within a type definition in the authorization model. It specifies the *possibility* of a relationship existing between an object of that type and a user. Relation names are arbitrary (e.g., owner, editor, viewer, member, admin), but must be defined on the object type in the model.
```

The following table summarizes these core concepts:

| Concept | Description | Example |
| :---- | :---- | :---- |
| **Type** | A class of objects that share similar characteristics. | user, document, folder, organization |
| **Object** | A specific instance of a defined type, an entity in the system. | `user:anne`, `document:report_2023`, `folder:marketing_docs` |
| **User** | An entity that can be related to an object. Can be a specific object, a wildcard, or a userset. | `user:bob`, `user:*` (everyone), `team:product#member` |
| **Relation** | A string defined within a type definition that specifies the possibility of a relationship between an object of that type and a user. | owner, editor, viewer, member, admin |
| **Relationship Tuple** | A grouping of a user, a relation, and an object, representing a factual relationship in the system. | `{"user": "user:anne", "relation": "viewer", "object": "document:roadmap"}` |
| **Authorization Model** | A static definition combining type definitions to define the entire permission structure of a system. | ```type document relations define viewer: [user]``` |

## **3. The OpenFGA Modeling Language: DSL**

> OpenFGA's Configuration Language is fundamental to constructing a system's authorization model, informing OpenFGA about object types and their relationships.10 It describes all possible relations for an object of a given type and the conditions under which one entity is related to that object.10 The language is primarily expressed in DSL (Domain Specific Language).

Reviewer: ```suggestion
OpenFGA's Configuration Language is fundamental to constructing a system's authorization model, informing OpenFGA about object types and their relationships. It describes all possible relations for an object of a given type and the conditions under which one entity is related to that object. The language is primarily expressed in DSL (Domain Specific Language).
```

### **DSL: Developer-Friendly Syntax for Readability**

The DSL provides syntactic sugar over the underlying JSON, designed for ease of use and improved readability.10 It is the preferred syntax for developers using the Playground, CLI, and IDE extensions (like Visual Studio Code), offering features like syntax highlighting and validation.5 DSL models are compiled to JSON before being sent to OpenFGA's API.5

An example of an OpenFGA model in DSL:

```dsl.openfga
model  
  schema 1.1  
type user  
type document  
  relations  
  define viewer: [user] or editor  
  define editor: [user]
```

## **4  Defining Relationships: Crafting Your Authorization Logic**

OpenFGA provides a rich set of constructs for defining relationships, enabling the modeling of complex authorization policies.

### **Direct Relationships: Explicit Access Grants**

> A direct relationship is established when a specific relationship tuple (e.g., user=X, relation=R, object=Y) is explicitly stored. The authorization model must explicitly permit this through direct relationship type restrictions.5 These restrictions define which types of users can be directly associated with an object for a given relation, using formats like

Reviewer: ```suggestion
A direct relationship is established when a specific relationship tuple (e.g., user=X, relation=R, object=Y) is explicitly stored. The authorization model must explicitly permit this through direct relationship type restrictions. These restrictions define which types of users can be directly associated with an object for a given relation, using formats like
```

`[<type>`], `[<type:*>]`, or `[<type\>#<relation\>]`. 

For example,

```
define owner: [user] 
```

means only individual users can be directly assigned as owners.

A tuple like:

```

> {"user": "user:anne", "relation": "user", "object": "document:1"} 

Reviewer: ```suggestion
{"user": "user:anne", "relation": "owner", "object": "document:1"} 
```

```

is a direct relationship.

### **Concentric Relationships: Inheriting Permissions**

Concentric relationships represent nested or implied relations, where one relation automatically confers another (e.g., "all editors are viewers").This is implemented using the or keyword within a relation definition.

For example,

```
define viewer: [user] or editor
```

> means a user is a viewer if directly assigned OR if they are an editor. user:anne is an editor of `document:new-roadmap`, she implicitly has viewer access, reducing the number of required tuples.

Reviewer: ```suggestion
means a user is a viewer if directly assigned OR if they are an editor. `user:anne` is an editor of `document:new-roadmap`, she implicitly has viewer access, reducing the number of required tuples.
```

### **Indirect Relationships with 'X from Y': Scalable Hierarchical and Group-Based Access**

The X from Y syntax is crucial for scalability, allowing a user to acquire a relation (X) to an object through an intermediary object (Y) and a defined relation on Y. This avoids individual tuple creation for every permission, enabling higher-level abstraction. It is highly effective for hierarchical and group-based access control. For example,

```
define admin: [user] or repo_admin from organization 
```

on a repo type means a user is an admin if directly assigned, or if they have the repo_admin relation to an org that owns the repo.

This simplifies management; revoking access can be done by deleting a single tuple linking the intermediary.

### **Contextual Authorization with Conditions: Dynamic Permissions**

Conditions introduce dynamic, contextual authorization. A condition is a function using Google's Common Expression Language (CEL), with parameters and a boolean expression.

Example:
```
condition less_than_hundred(x: int) { 
  x < 100 
}
```

> Conditions are need to be added at the end of the model, and are instantiated using conditional relationship tuples.

Reviewer: ```suggestion
Conditions are required to be defined at the end of the model (after the type definitions), and are instantiated using conditional relationship tuples.
```

```dsl.openfga
type folder  
  relations  
    define parent_folder: [folder] 
    define owner: [user]  
    define editor: [user] or owner or editor from parent_folder  
    define viewer: [user] or editor or viewer from parent_folder  

    define can_view: viewer
    define can_edit: editor
    define can_delete: editor
    define can_share: owner
```

Always define permissions in the authorization models.

## **7. Testing and Validating Your OpenFGA Models**

Thorough testing and validation are indispensable. OpenFGA provides tools for rapid prototyping and automated testing to ensure your authorization model is correctly designed before deployment.

### **OpenFGA CLI: Command-Line Model Management and Testing**

The OpenFGA CLI serves as a cross-platform command-line tool for interacting with an OpenFGA server. It provides a robust set of commands for various tasks, including:

* Reading, writing, validating, and transforming authorization models.
* Running tests on an authorization model, which is crucial for verifying its correctness.
* Managing OpenFGA stores, encompassing operations such as creation, listing, retrieval, deletion, import, and expor

> To execute tests defined within a .fga.yaml file, the fga model test command is utilized: fga model test \--tests \<filename\>.fga.yaml.18

Reviewer: ```suggestion
To execute tests defined within a .fga.yaml file, the fga model test command is utilized: `fga model test --tests <filename>.fga.yaml`.
```

### **Automated Testing with .fga.yaml**

The .fga.yaml file is central to defining and testing OpenFGA authorization models, providing a structured approach for validation.18 This file can include:

> * name (optional): A descriptive name for the test file.18  

Reviewer: ```suggestion
* `name` (optional): A descriptive name for the test file.
```

> * model or model_file: The authorization model can be defined inline or referenced from an external .fga, .json, or .mod file.18  

Reviewer: ```suggestion
* `model` or `model_file`: The authorization model can be defined inline or referenced from an external .fga, .json, or .mod file.
```

> * tuples or tuple_file or tuple_files (optional): Relationship tuples can be defined inline or referenced from external JSON, YAML, or CSV files, and are considered for all tests.18

Reviewer: ```suggestion
* `tuples` or `tuple_file` or `tuple_files` (optional): Relationship tuples can be defined inline or referenced from external JSON, JSONL, YAML, or CSV files, and are considered for all tests.
```

#### Defining the Model and Tuples in .fga.yaml  

You can define your authorization model and relationship tuples directly within the .fga.yaml file. This allows for a self-contained test definition.  

> Example of defining a model and tuples inline 18:

Reviewer: ```suggestion
Example of defining a model and tuples inline:
```

  }  
tuples: # Inline tuple definitions go here  
  # Anne is a member of the Acme organization  
  - user: user:anne  
    relation: member  
    object: organization:acme  
  # Peter has the admin role from February 2nd 2024 0AM to 1AM  
  - user: user:peter  
    relation: admin  
    object: organization:acme  
    condition:  
      name: non_expired_grant  
      context:  
        grant_time : "2024-02-01T00:00:00Z"  
        grant_duration : 1h

```

#### Writing Tests in .fga.yaml  

The tests section within the .fga.yaml file is where you define specific test cases for OpenFGA API calls, including check, list_objects, and list_users. 

* **Check Tests:** Verify whether a user U has a relation R with an object O. These tests include the user, object, context (for evaluating conditions), and assertions (pairs of relation: expected-result).

  Example of Check tests 
```yaml
  tests:  
    - name: Test  
      check:  
        - user: user:peter  
          object: organization:acme  
          context:  
            current_time : "2024-02-01T00:10:00Z"  
          assertions:  
            member: false  
            admin: true
```

> * **List Objects Tests:** Validate the expected results when querying which objects a user has a specific relation with (e.g., user:anne is a member of organization:acme). These tests include the user, type of object, context, and assertions.18  Example of List Objects tests:

Reviewer: ```suggestion
* **List Objects Tests:** Validate the expected results when querying which objects a user has a specific relation with (e.g., user:anne is a member of organization:acme). These tests include the user, type of object, context, and assertions.  Example of List Objects tests:
```
